examples/nemo/eg5/extract_kernels.py

```diff
@@ -0,0 +1,105 @@
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors: R. W. Ford, A. R. Porter, S. Siso and N. Nobre, STFC Daresbury Lab
+
+'''A transformation script that seeks to apply OpenACC DATA and KERNELS
+directives to NEMO style code.  In order to use
+it you must first install PSyclone. See README.md in the top-level
+psyclone directory.
+
+Once you have psyclone installed, this may be used by doing:
+
+ $ psyclone -api nemo -s ./kernels_trans.py some_source_file.f90
+
+This should produce a lot of output, ending with generated
+Fortran. Note that the Fortran source files provided to PSyclone must
+have already been preprocessed (if required).
+
+The transformation script attempts to insert Kernels directives at the
+highest possible location(s) in the schedule tree (i.e. to enclose as
+much code as possible in each Kernels region). However, due to
+limitations in the PGI compiler, we must take care to exclude certain
+nodes (such as If blocks) from within Kernel regions. If a proposed
+region is found to contain such a node (by the ``valid_kernel``
+routine) then the script moves a level down the tree and then repeats
+the process of attempting to create the largest possible Kernel
+region.
+
+Once the Kernels regions have been created, the script then simply
+encloses each of them within an OpenACC Data region (since these have
+already been made as large as possible). In reality, the purpose of a
+data region is to keep data on the remote GPU device for as long as
+possible, ideally between Kernel regions. However, this requires more
+sophisticated dependency analysis than is yet implemented in
+PSyclone. Issue #309 will tackle this limitation.
+
+'''
+
+from psyclone.nemo import NemoLoop
+from psyclone.transformations import TransformationError
+from psyclone.psyir.transformations import ExtractTrans
+
+
+def trans(psy):
+    '''A PSyclone-script compliant transformation function. Applies
+    OpenACC 'kernels' and 'data' directives to NEMO code.
+
+    :param psy: The PSy layer object to apply transformations to.
+    :type psy: :py:class:`psyclone.psyGen.PSy`
+    '''
+
+    extract = ExtractTrans()
+
+    print("Invokes found:\n" +
+          "\n".join([str(name) for name in psy.invokes.names]) + "\n")
+
+    for invoke in psy.invokes.invoke_list:
+
+        sched = invoke.schedule
+        if not sched:
+            print(f"Invoke {invoke.name} has no Schedule! Skipping...")
+            continue
+
+        count = 0
+        for kern in sched.children:
+            count += 1
+            if not isinstance(kern, NemoLoop):
+                continue
+            try:
+                extract.apply(kern,
+                              options={"COLLECT-ARRAY-SHAPE-READS": False})
```

Comment:
Please add a comment explaining what the "COLLECT-ARRAY-SHAPE-READS" option setting is for.

Reply:
Done - but tbh, we might need to look at a better option here.  FWIW, reason is the expression:
```
zwx(:,:,jpk) = 0.e0
```
which (I assume) is lowered to an expression using `shape(zwx, ..)` on the RHS, which then gives a problem with a combined read/write access on the left hand side. Of course, jf we should have code lik:
```
something_else(shape(zwx, 1),jpk) = 0.e0
```
then we do need the read access (or we get smart and only store the result of `shape`. But that will require a lot of additional work in the driver later, which I wouldn't like ... )? I'd guess converting array expressions to explicit loops would also solve this issue.